### ISSUE TITLE
Message documentation vs production

### ISSUE THREAD

**Issue by pypingou** [CentOS-PaaS-SIG/ci-pipeline#135]
For the fedmsg client of ccsdb (http://resultsdb.ci.centos.org/ccsdb/) I based myself on the documentated format for the fedmsg messages present in the readme file [1] of this project.
I am now seeing some messages being published that differ quite a bit from the announce format, so I am curious if the current format is temporary or final. If the later then we should adjust our documentation.

Expected:
````
{ 
  u'i': 1,
  u'msg': { u'build_url': u'<full-jenkins-url-to-build>',
            u'package_url': u'<full-url-to-package>',
            u'build_id': <Jenkins-build-id-number>,
            u'ref': u'fedora/rawhide/${basearch}/atomic-host',
      	u'rev': u'<sha of the commit from dist-git>', 
      	u'namespace': u'rpms', 
      	u'repo': u'php-simplepie',
            u'status': u'<success>',
            u'test_guidance': u'<comma-separated-list-of-test-suites-to-run>'},
  u'msg_id': u'2015-4b5aae66-b713-4c22-bb4a-1277d4402375',
  u'timestamp': 1379638157.759283,
  u'topic': u'org.centos.prod.ci.pipeline.package.complete',
  u'username': u'fedora-atomic'
}
````

Observed:
````
{
  "i": 1, 
  "timestamp": 1496243297, 
  "msg_id": "2017-71d0a514-7925-4e62-9a92-19c1ebdc352d", 
  "crypto": "x509", 
  "topic": "org.centos.prod.ci.pipeline.package", 
  "msg": {
    "CI_TYPE": "custom", 
    "username": "fedora-atomic", 
    "message-content": "{\n  \"build_url\": \"https://ci.centos.org/job/ci-pipeline-rpmbuild-trigger/17354/\",\n  \"build_id\": \"17354\",\n  \"ref\": \"fedora/f25/x86_64/atomic-host\",\n  \"rev\": \"578ea71331a63e67b8681c02638af400d76d6611\",\n  \"namespace\": \"rpms\",\n  \"repo\": \"krita\",\n  \"status\": \"success\",\n  \"test_guidance\": \"\"\n}", 
    "CI_NAME": "ci-pipeline-rpmbuild-trigger", 
    "topic": "org.centos.prod.ci.pipeline.package.ignore", 
    "CI_STATUS": "passed"
  }
}
````
See that the topic is not consistent in the outer envelop versus the inner one, the CI_TYPE, CI_NAME, CI_STATUS and messages-content fields are unexpected.

I am fine with adjusting my code but I just would like to know if I should or if the format going to be changed to fit the documentation.

Thanks!

[1] https://github.com/CentOS-PaaS-SIG/ci-pipeline/#message-types

**Comment by johnbieren** [comment]
The reason why the topic is not consistent in the message you pasted is because the jms messaging plugin was recently updated to allow topic to be a variable in both override-topic and in the message properties, but that is not yet live in ci.centos.org, so the override-topic version of topic had to be cut off slightly for now until that plugin is pushed into production for ci.centos.  Only pipeline.package messages have this inconsistency currently, as far as I know.

**Comment by arilivigni** [comment]
@pypingou we will look into this with @scoheb to see how we can formalize the format if it needs to be or update our documentation.  The key things are in the msg: key.  I suspect we have to update our spec since the JMS messaging plugin wraps things.

**Comment by pypingou** [comment]
I'm fine either way, just wanted us to be consistent and know what is the stable version :)

(I do have a preference to no serializing twice, but it's not a big deal if I have to deserialize myself).

**Comment by arilivigni** [comment]
@pypingou glad you raised it since we want to be consistent

**Comment by arilivigni** [comment]
@pypingou I am going to update the documentation to unblock you with what we are currently producing even though it really isn't that much different.

**Comment by arilivigni** [comment]
@johnbieren @p3ck would it be better if we just put all key/value pairs in the message properties and not content?  Then we are not wrapping things again.
I did this in the pipeline and the output is a much closer match and I added the branch as a separate field.  Also I noticed there is no url link to the package, but I assume we can add that.

{
  "i": 1, 
  "timestamp": 1496387649, 
  "msg_id": "2017-18548f1e-7146-4bc0-a968-8ad07015361e", 
  "crypto": "x509", 
  "topic": "org.centos.prod.ci.pipeline.package.complete", 
  "msg": {
    "CI_TYPE": "custom", 
    "build_id": "79", 
    "username": "fedora-atomic", 
    "rev": "0546fa18041a8ca223f4f441dc1868fc81ddce0f", 
    "message-content": "", 
    "build_url": "https://ci.centos.org/job/ci-pipeline-duffy-allocate-new/79/", 
    "namespace": "rpms", 
    "CI_NAME": "ci-pipeline-duffy-allocate-new", 
    "repo": "perl", 
    "topic": "org.centos.prod.ci.pipeline.package.complete", 
    "status": "success", 
    "branch": "rawhide", 
    "test_guidance": "''", 
    "ref": "fedora/rawhide/x86_64/atomic-host"
  }
}

@pypingou Does this make sense now?

**Comment by johnbieren** [comment]
@arilivigni +1 for putting the key value pairs in the properties instead.  Based on how the messages are being received that seems like a good way to fix it to me.

**Comment by scoheb** [comment]
+1 Nice and clean solution.

**Comment by pypingou** [comment]
Looks good to me as well, thanks for including the branch in its own element while at it :)

**Comment by arilivigni** [comment]
Great I am going to adjust the documents to reflect it.

**Comment by johnbieren** [comment]
The jenkins jobs should be fixed by
https://github.com/CentOS-PaaS-SIG/ci-pipeline/pull/144
and
https://github.com/CentOS-PaaS-SIG/ci-pipeline/pull/145

**Comment by pypingou** [comment]
Awesome! :)

I think all we need now is to adjust the README.

Is there an ETA for this in prod? (just for me to know when I should expect my client to fail)

**Comment by pypingou** [comment]
> Is there an ETA for this in prod? (just for me to know when I should expect my client to fail)

Lol:
````
[2017-06-02 14:44:49][    fedmsg    INFO] 
{
  "i": 1, 
  "timestamp": 1496414697, 
  "msg_id": "2017-f8ef696d-e2f9-41a2-84de-991a5644ac04", 
  "crypto": "x509", 
  "topic": "org.centos.prod.ci.pipeline.compose.running", 
  "msg": {
    "CI_TYPE": "custom", 
    "build_id": "494", 
    "username": "fedora-atomic", 
    "compose_url": "http://artifacts.ci.centos.org/artifacts/fedora-atomic/rawhide/ostree", 
    "rev": "2aa8a11e2c2962844e4081c2f1994975d1b96aed", 
    "message-content": "", 
    "build_url": "https://ci.centos.org/job/ci-pipeline-ostree-compose/494/", 
    "namespace": "rpms", 
    "CI_NAME": "ci-pipeline-ostree-compose", 
    "repo": "libgcrypt", 
    "compose_rev": "", 
    "topic": "org.centos.prod.ci.pipeline.compose.running", 
    "status": "success", 
    "branch": "rawhide", 
    "test_guidance": "", 
    "ref": "fedora/rawhide/x86_64/atomic-host"
  }
}
````
I guess that answers my question :)

**Comment by pypingou** [comment]
FTR, this is all it took to make the fedmsg consumer work with the published message https://pagure.io/ccsdb/c/2aeb240e3aea2e041031aefbf35820ced12268b0?branch=master (vs the documented ones)

**Comment by scoheb** [comment]
But topic is still a top level attribute...and it guaranteed to be there.
The inner topic is added by the user.

-         topic = msg['topic']
+         topic = msg['msg']['topic']

Can you not keep it as it was?

**Comment by pypingou** [comment]
> Can you not keep it as it was?

Not until the bug that truncates the outer topic is fixed, after that, I'd be happy to :)

(Or then we could just entirely drop the inner topic, is it really useful to keep it there twice?)

**Comment by arilivigni** [comment]
@p3ck @johnbieren @pypingou 

Please make sure the update in the document makes sense for what we agreed for the format:
https://github.com/CentOS-PaaS-SIG/ci-pipeline/pull/146

**Comment by arilivigni** [comment]
This should be resolved since PR #146 was merged
